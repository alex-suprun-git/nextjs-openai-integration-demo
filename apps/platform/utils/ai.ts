import { getLocale } from 'next-intl/server';
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import { OutputFixingParser, StructuredOutputParser } from 'langchain/output_parsers';
import { PromptTemplate } from '@langchain/core/prompts';
import { Document } from 'langchain/document';
import { loadQARefineChain } from 'langchain/chains';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import z from 'zod';

const modelParams = {
  model: 'o4-mini',
};

const parser = StructuredOutputParser.fromZodSchema(
  z.object({
    mood: z.string().describe('the mood of the person who wrote the journal entry.'),
    subject: z.string().describe('the subject of the journal entry.'),
    negative: z
      .boolean()
      .describe('is the journal entry negative? (i.e. does it contain negative emotions?).'),
    summary: z.string().max(100).describe('quick summary of the entire entry.'),
    color: z
      .string()
      .describe(
        'a hexidecimal color code that represents the mood of the entry. Example #0101fe for blue representing happiness.',
      ),
    sentimentScore: z
      .number()
      .describe(
        'sentiment of the text and rated on a scale from -10 to 10, where -10 is extremely negative, 0 is neutral, and 10 is extremely positive.',
      ),
    title: z.string().describe('A concise title summarizing the journal entry.'),
  }) as any,
);

const getPrompt = async (content: string, language: UserLocale) => {
  const formatInstructions = parser.getFormatInstructions();

  const templates = {
    de: 'Analysiere den folgenden Tagebucheintrag und antworte auf Deutsch. Übersetze die Formatierungsanweisungen ins Deutsche und befolge sie unbedingt! \n{formatInstructions}\n{entry}',
    en: 'Analyze the following journal entry and respond in English. Follow the instructions and format your response to match the format instructions, no matter what! \n{formatInstructions}\n{entry}',
  };

  const prompt = new PromptTemplate({
    template: templates[language],
    inputVariables: ['entry'],
    partialVariables: { formatInstructions },
  });

  const input = await prompt.format({
    entry: content,
  });

  return input;
};

export const analyzeEntry = async (content: string) => {
  const locale = await getLocale();

  const input = await getPrompt(content, locale as UserLocale);
  const model = new ChatOpenAI({
    ...modelParams,
    apiKey: process.env.OPENAI_API_KEY,
  });
  const output = await model.invoke(input);

  try {
    return await parser.parse(output.content as string);
  } catch (e) {
    const fixParser = OutputFixingParser.fromLLM(new ChatOpenAI(modelParams), parser);
    const fix = await fixParser.parse(output.content as string);
    return fix;
  }
};

export const analysisFeedback = async (
  language: UserLocale,
  question: string,
  entries: BaseEntry[],
) => {
  if (!entries.length) {
    return undefined;
  }

  const docs = entries.map(
    (entry) =>
      new Document({
        pageContent: entry.content as string,
        metadata: { source: entry.id, date: entry.createdAt },
      }),
  );
  const model = new ChatOpenAI(modelParams);
  const chain = loadQARefineChain(model);
  const embeddings = new OpenAIEmbeddings();
  const store = await MemoryVectorStore.fromDocuments(docs, embeddings);
  const relevantDocs = await store.similaritySearch(question);

  const templates = {
    en: `Answer the following question focusing exclusively on the person's mood and feelings based on the given documents. If the question touches on any other topic or if the context is unclear, please respond with a polite request to rephrase the question. Ensure your answer is in the same language as the input. \nQuestion: {question}`,
    de: `Beantworte die folgende Frage ausschließlich im Hinblick auf die Stimmung und Gefühle der Person basierend auf den gegebenen Dokumenten. Falls die Frage ein anderes Thema betrifft oder der Kontext unklar ist, bitte höflich um eine Umformulierung der Frage. Stelle sicher, dass deine Antwort in derselben Sprache wie die Eingabe erfolgt. \nFrage: {question}`,
  };

  const promptTemplate = new PromptTemplate({
    template: templates[language],
    inputVariables: ['question'],
  });

  const input = await promptTemplate.format({
    question: question,
  });

  const res = await chain.invoke({
    input_documents: relevantDocs,
    question: input,
  });

  return res.output_text;
};
