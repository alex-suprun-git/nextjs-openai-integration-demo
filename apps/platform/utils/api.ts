export const createUrl = (path: string) => window.location.origin + path;

export const updateUserPromptUsage = async (promptContentLength: number) => {
  const res = await fetch(
    new Request('/api/user', {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ promptContentLength }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const createNewEntry = async (content: string) => {
  const res = await fetch(
    new Request('/api/journal', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

// @Deprecated
export const updateEntry = async (id: string, content: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const deleteEntry = async (id: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  return;
};

export const askQuestion = async (question: string) => {
  const res = await fetch(
    new Request('/api/question', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ question }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data;
};
