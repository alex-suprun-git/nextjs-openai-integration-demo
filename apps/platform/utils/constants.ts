export const MINIMUM_CONTENT_LENGTH = 30;
export const COOKIE_NAME = 'NEXT_LOCALE';
export const ENTRIES_BASE_PATH = '/records';
