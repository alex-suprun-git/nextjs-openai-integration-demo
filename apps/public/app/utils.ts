export const createUrl = (path: string) => window.location.origin + path;

export const getPageURL = (url: string) => {
  const [, , ...restSegments] = url.split('/');
  return restSegments.join('/');
};

export async function fetcher(_url: string) {
  const url = createUrl(_url);

  const res = await fetch(
    new Request(url, {
      method: 'GET',
    }),
  );

  if (!res.ok) return;
  return res.json() as Promise<{ isSignedIn: boolean }>;
}
