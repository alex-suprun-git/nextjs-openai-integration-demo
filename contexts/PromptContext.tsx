'use client';

import React, { createContext, useContext, ReactNode } from 'react';

interface PromptContextProps {
  promptSymbolsLimit: number;
  promptSymbolsUsed: number;
}

const PromptContext = createContext<PromptContextProps | undefined>(undefined);

export const usePrompt = () => {
  const context = useContext(PromptContext);
  if (!context) {
    throw new Error('usePrompt must be used within a PromptProvider');
  }
  return context;
};

interface PromptProviderProps {
  children: ReactNode;
  value: PromptContextProps;
}

export const PromptProvider: React.FC<PromptProviderProps> = ({ children, value }) => (
  <PromptContext.Provider value={value}>{children}</PromptContext.Provider>
);
