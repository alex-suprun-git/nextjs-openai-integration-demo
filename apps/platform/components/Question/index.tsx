'use client';

import { FormEvent, useState } from 'react';
import { useTranslations } from 'next-intl';
import { HiMiniChevronDoubleRight } from 'react-icons/hi2';
import { askQuestion, updateUserPromptUsage } from '@/utils/api';
import { usePrompt } from '@/contexts/PromptContext';
import { Loading } from '@/ui-lib';

const Question = () => {
  const [question, setQuestion] = useState('');
  const [answer, setAnswer] = useState<string | null>(null);
  const [isLoading, setLoading] = useState(false);

  const { symbolsUsed, symbolsLimit } = usePrompt();
  const isPromptSymbolsExceeded = +symbolsUsed >= +symbolsLimit;

  const t = useTranslations('analysisRequest');

  if (isPromptSymbolsExceeded) {
    return null;
  }

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setLoading(true);

    try {
      const { data } = await askQuestion(question);
      setAnswer(data);
      await updateUserPromptUsage(question.length);
      setQuestion('');
    } catch (error) {
      setAnswer(t('labels.responseError'));
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="my-12">
      <form data-testid="promptInput-wrapper" onSubmit={handleSubmit}>
        <span className="mb-4 block font-semibold text-stone-300">{t('headline')}</span>
        <div className="flex">
          <input
            id="analysis-request"
            type="text"
            value={question}
            onChange={(e) => setQuestion(e.target.value)}
            className="mr-2 w-[100%] rounded-md border border-gray-300 bg-slate-700 p-2 text-stone-300 outline-none md:w-[50%] lg:w-[33%]"
            disabled={isLoading}
            minLength={15}
            maxLength={100}
            placeholder={t('labels.inputPlaceholder')}
          />
          {question && (
            <button
              disabled={isLoading}
              type="submit"
              className="btn border-0 border-gray-300 px-4 py-2 text-xl disabled:cursor-not-allowed disabled:bg-gray-300 disabled:text-black"
            >
              <HiMiniChevronDoubleRight />
            </button>
          )}
        </div>
      </form>
      {isLoading && <Loading fullscreen />}
      {!isLoading && answer && (
        <p data-testid="promptInput-answer" className="text my-4 md:max-w-[50%]">
          {answer}
        </p>
      )}
    </div>
  );
};

export default Question;
