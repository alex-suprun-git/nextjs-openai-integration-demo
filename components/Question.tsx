'use client';

import { HiMiniChevronDoubleRight } from 'react-icons/hi2';
import { FormEvent, useState } from 'react';
import { askQuestion, updateUser } from '@/utils/api';
import { usePrompt } from '@/contexts/PromptContext';
import { useRouter } from 'next/navigation';

const Question = () => {
  const [question, setQuestion] = useState('');
  const [answer, setAnswer] = useState(null);
  const [isLoading, setLoading] = useState(false);
  const router = useRouter();

  const { promptSymbolsUsed, promptSymbolsLimit } = usePrompt();
  const isPromptSymbolsExceeded = promptSymbolsUsed >= promptSymbolsLimit;

  if (isPromptSymbolsExceeded) {
    return null;
  }

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setLoading(true);

    const { data } = await askQuestion(question);

    setAnswer(data);
    setLoading(false);
    await updateUser(promptSymbolsUsed + question.length);
    setQuestion('');
    router.refresh();
  };

  return (
    <div className="my-12">
      <form onSubmit={handleSubmit}>
        <span className="mb-4 block font-semibold text-white">
          Request an AI analysis of your notes (10 to 100 characters)
        </span>
        <div className="flex">
          <input
            id="analysis-request"
            type="text"
            value={question}
            onChange={(e) => setQuestion(e.target.value)}
            className="mr-2 w-[100%] rounded-md border border-gray-300 bg-slate-700 p-2 outline-none md:w-[50%] lg:w-[33%]"
            disabled={isLoading}
            minLength={10}
            maxLength={100}
            placeholder="e.g. how good was my week in average?"
          />
          {question && (
            <button
              disabled={isLoading}
              type="submit"
              className="btn border-gray-300 px-4 py-2 text-xl disabled:cursor-not-allowed disabled:bg-gray-300 disabled:text-black"
            >
              <HiMiniChevronDoubleRight />
            </button>
          )}
        </div>
      </form>
      {isLoading && <div className="loading relative top-5"></div>}
      {!isLoading && answer && <p className="text my-4 md:max-w-[50%]">{answer}</p>}
    </div>
  );
};

export default Question;
