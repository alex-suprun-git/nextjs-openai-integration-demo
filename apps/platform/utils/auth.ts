import { getServerSession } from 'next-auth';
import { authOptions } from '@/app/api/auth/[...nextauth]/auth';
import { prisma } from './db';
import { Prisma } from '@prisma/client';
import { isDynamicServerError } from 'next/dist/client/components/hooks-server-context';

export const getCurrentUser = async () => {
  try {
    const session = await getServerSession(authOptions);

    if (!session?.user?.email) {
      console.warn('User email is not available. User might be signed out.');
      return null;
    }

    const user = await prisma.user.findFirstOrThrow({
      where: {
        email: session.user.email as string,
      },
    });

    return user;
  } catch (error: any) {
    if (error instanceof Prisma.PrismaClientValidationError) {
      console.error('Prisma validation error:', error.message);
      throw new Error('Failed to find user due to validation error.');
    } else if (error.code === 'P2025') {
      console.error('User not found:', error.message);
      throw new Error('User not found.');
    } else {
      if (isDynamicServerError(error)) {
        throw error;
      }
      console.error('Unexpected error:', error);
      throw new Error('An unexpected error occurred.');
    }
  }
};
