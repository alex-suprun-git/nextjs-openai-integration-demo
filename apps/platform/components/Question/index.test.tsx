import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, Mock } from 'vitest';
import { createTranslator, useTranslations } from 'next-intl';
import Question from '.';
import { askQuestion, updateUserPromptUsage } from '@/utils/api';
import { usePrompt } from '@/contexts/PromptContext';
import { useRouter } from 'next/navigation';

vi.mock('next/navigation', () => ({
  useRouter: vi.fn(),
}));

vi.mock('@/utils/api', () => ({
  askQuestion: vi.fn(),
  updateUserPromptUsage: vi.fn(),
}));

vi.mock('@/contexts/PromptContext', () => ({
  usePrompt: vi.fn(),
}));

describe('Question', () => {
  const getPlaceholderRegex = /e.g. how good was my week in average?/i;

  const mockRefresh = vi.fn();

  const mockPromptContext = {
    symbolsUsed: '50',
    symbolsLimit: '100',
  };

  beforeAll(async () => {
    const translate = createTranslator({
      locale: 'en',
      namespace: 'analysisRequest',
      messages: (await import('@/messages/en.json')).default,
    });

    (useTranslations as Mock).mockImplementation(() => translate);
  });

  beforeEach(() => {
    vi.clearAllMocks();
    (usePrompt as Mock).mockReturnValue(mockPromptContext);

    (useRouter as Mock).mockReturnValue({
      refresh: mockRefresh,
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(<Question />);
    expect(container).toMatchSnapshot();
  });

  it('disables input and button when loading', async () => {
    (askQuestion as Mock).mockResolvedValueOnce({ data: 'This is the answer' });

    render(<Question />);
    fireEvent.change(screen.getByPlaceholderText(getPlaceholderRegex), {
      target: { value: 'This is a test question' },
    });

    const button = await screen.findByRole('button');

    fireEvent.submit(button);

    await waitFor(() => {
      expect(screen.getByPlaceholderText(getPlaceholderRegex)).toBeDisabled();
    });

    await waitFor(() => {
      expect(button).toBeDisabled();
    });

    await waitFor(() => {
      expect(screen.getByText('This is the answer')).toBeInTheDocument();
    });
  });

  it('calls askQuestion and updateUserPromptUsage on submit', async () => {
    (askQuestion as Mock).mockResolvedValueOnce({ data: 'This is the answer' });

    render(<Question />);

    fireEvent.change(screen.getByPlaceholderText(getPlaceholderRegex), {
      target: { value: 'This is a test question' },
    });

    fireEvent.submit(screen.getByRole('button'));

    await waitFor(() => {
      expect(askQuestion).toHaveBeenCalledWith('This is a test question');
    });

    await waitFor(() => {
      expect(updateUserPromptUsage).toHaveBeenCalledWith(23); // content length
    });

    await waitFor(() => {
      expect(screen.getByText('This is the answer')).toBeInTheDocument();
    });
  });

  it('does not render when prompt symbols are exceeded', () => {
    (usePrompt as Mock).mockReturnValue({
      symbolsUsed: '100',
      symbolsLimit: '100',
    });

    const { container } = render(<Question />);
    expect(container).toBeEmptyDOMElement();
  });

  it('shows an error message on failed question submission', async () => {
    (askQuestion as Mock).mockRejectedValueOnce(new Error('Failed to fetch'));

    render(<Question />);

    fireEvent.change(screen.getByPlaceholderText(getPlaceholderRegex), {
      target: { value: 'This is a test question' },
    });

    fireEvent.submit(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByText('Error occurred')).toBeInTheDocument();
    });
  });
});
