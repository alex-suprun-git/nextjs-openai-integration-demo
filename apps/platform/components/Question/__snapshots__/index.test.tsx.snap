// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Question > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="my-12"
  >
    <form
      data-testid="promptInput-wrapper"
    >
      <span
        class="mb-4 block font-semibold text-stone-300"
      >
        Request an AI analysis of your notes (15 to 100 characters)
      </span>
      <div
        class="flex"
      >
        <input
          class="mr-2 w-[100%] rounded-md border border-gray-300 bg-slate-700 p-2 text-stone-300 outline-none md:w-[50%] lg:w-[33%]"
          id="analysis-request"
          maxlength="100"
          minlength="15"
          placeholder="e.g. how good was my week in average?"
          type="text"
          value=""
        />
      </div>
    </form>
  </div>
</div>
`;
